numpy/core/shape_base.py

```diff
@@ -417,38 +417,68 @@ def format_index(index):
-        return parent_index + [None], 0
+        raise ValueError('Lists cannot be empty')
     else:
         # We've 'bottomed out' - arrays is either a scalar or an array
         return parent_index, _nx.ndim(arrays)
 
 
-def _block(arrays, max_depth, result_ndim):
+def _concatenate_shapes(shapes, axis):
+    """Given array shapes, return the resulting shape that would occur
+    after array concatenation.
+
+    concatenate(arrs, axis).shape == _concatenate_shapes([a.shape for a in arrs], axis)
     """
-    Internal implementation of block. `arrays` is the argument passed to
-    block. `max_depth` is the depth of nested lists within `arrays` and
-    `result_ndim` is the greatest of the dimensions of the arrays in
-    `arrays` and the depth of the lists in `arrays` (see block docstring
-    for details).
+    # Take a shape, any shape
+    first_shape = shapes[0]
+    for shape in shapes[1:]:
+        if (shape[:axis] != first_shape[:axis] or
+                shape[axis+1:] != first_shape[axis+1:]):
+            raise ValueError('Mismatched array shapes in block.')
+    shape_on_dim = sum(shape[axis] for shape in shapes)
+    return first_shape[:axis] + (shape_on_dim,) + first_shape[axis+1:]
+
+
+def _atleast_nd(a, ndim):
+    # Ensures `a` has at least `ndim` dimensions by prepending
+    # ones to `a.shape` as necessary
+    return array(a, ndmin=ndim, copy=False, subok=True)
+
+
+def _block_info_recursion(arrays, list_ndim, result_ndim,
+                          start_indices, depth=0):
+    """From the input arrays to block, return a tuple containing
+
+    * the resulting shape of the final array
+    * the resulting dtype of the final array
+    * A flat list of tuples containing pairs of (slice, arrays)
+      useful for assignment
     """
-    def atleast_nd(a, ndim):
-        # Ensures `a` has at least `ndim` dimensions by prepending
-        # ones to `a.shape` as necessary
-        return array(a, ndmin=ndim, copy=False, subok=True)
-
-    @recursive
-    def block_recursion(self, arrays, depth=0):
-        if depth < max_depth:
-            if len(arrays) == 0:
-                raise ValueError('Lists cannot be empty')
-            arrs = [self(arr, depth+1) for arr in arrays]
-            return _nx.concatenate(arrs, axis=-(max_depth-depth))
-        else:
-            # We've 'bottomed out' - arrays is either a scalar or an array
-            # type(arrays) is not list
-            return atleast_nd(arrays, result_ndim)
-
-    return block_recursion(arrays)
+    if depth < list_ndim:
+        concat_depth = len(start_indices)
+        shapes = []
+        dtypes = []
+        regions = []
+        concat_position = 0
+        for arr in arrays:
+            shape, dtype, new_regions = _block_info_recursion(
+                arr, list_ndim, result_ndim,
+                start_indices + (concat_position,),
+                depth + 1)
+            concat_position += shape[concat_depth]
```

Comment:
I think I'd like to see the computation of `concat_position` in a helper `_concatenate_offsets` function or similar, which you can then `zip` with `arrays`

Reply:
Whats your workaround for [`itertools.accumulate`](https://docs.python.org/3.7/library/itertools.html#itertools.accumulate) with regards to 2.7 compatibility.

Reply:
Thanks for the reminder that `accumulate` exists in py3!

I was thinking something simple like
```
def _concatenate_offsets(shapes, axis):
    offset = 0
    offsets = []
    for shape in shape:
        offsets.append(offset)
        offset += shape[axis]
    return offsets
```

Reply:
man, i'm close, but I think i'm banging my head too much, see if you want to take a look at it. I'll figure out py.27 compatibility.